### PR TITLE
REF: avoid getattr pattern for diff_2d; use fused types

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/_libs/algos_common_helper.pxi.in [hunk 1]
```diff
@@ -4,32 +4,34 @@ Template for each `dtype` helper function using 1-d template
...
-          ('int8',  'int8_t',  'float32_t'),
-          ('int16', 'int16_t', 'float32_t'),
-          ('int32', 'int32_t', 'float64_t'),
-          ('int64', 'int64_t', 'float64_t')]
-
-def get_dispatch(dtypes):
-
-    for name, c_type, dest_type, in dtypes:
-        yield name, c_type, dest_type
+ctypedef fused diff_t:
```
Comment (jreback) [inline pandas/_libs/algos_common_helper.pxi.in]: isn’t this numeric? can we consolidate these later

Comment (jbrockmendel) [reply pandas/_libs/algos_common_helper.pxi.in]: i hope so

**Review by WillAyd — APPROVED** [review]
lgtm

File: pandas/_libs/algos_common_helper.pxi.in [hunk 2]
```diff
@@ -4,32 +4,34 @@ Template for each `dtype` helper function using 1-d template
...
-def diff_2d_{{name}}(ndarray[{{c_type}}, ndim=2] arr,
-                     ndarray[{{dest_type}}, ndim=2] out,
-                     Py_ssize_t periods, int axis):
+def diff_2d(ndarray[diff_t, ndim=2] arr,
+            ndarray[out_t, ndim=2] out,
+            Py_ssize_t periods, int axis):
+
+    # Disable for unsupported dtype combinations,
+    #  see https://github.com/cython/cython/issues/2646
+    if out_t is float32_t:
```
Comment (jreback) [inline pandas/_libs/algos_common_helper.pxi.in]: assume this a later followup

**Comment by jreback** [comment]
thanks

File: pandas/_libs/algos_common_helper.pxi.in [hunk 3]
```diff
@@ -4,32 +4,34 @@ Template for each `dtype` helper function using 1-d template
...
+        if not (diff_t is float32_t or diff_t is int8_t or diff_t is int16_t):
+            raise NotImplementedError
+    else:
+        if (diff_t is float32_t or diff_t is int8_t or diff_t is int16_t):
+            raise NotImplementedError
```
Comment (jbrockmendel) [inline pandas/_libs/algos_common_helper.pxi.in]: @WillAyd you may have noticed a bunch of build warnings about unreachable code have started showing up.  I'm pretty sure this is the reason.  Any thoughts on more graceful ways to get the combination of dtypes we're interested in?  (see the GH link on L27)

The same dtype-combination thing is the only reason why algos_take_helper isn't converted to fused types; it'd be really nice to clear that up

Comment (WillAyd) [reply pandas/_libs/algos_common_helper.pxi.in]: Don't know off the top of my head - maybe an Enum using the type as a member?

**Pull request merged**